libraries/chain/backing_store/db_key_value_format.cpp

```diff
@@ -16,37 +38,225 @@ namespace eosio { namespace chain { namespace backing_store { namespace db_key_v
-      b1::chain_kv::bytes composite_key;
-      detail::prepare_composite_key<uint64_t, key_type::primary>(composite_key, scope, table, sizeof(uint64_t));
-      b1::chain_kv::append_key(composite_key, db_key);
+   b1::chain_kv::bytes create_primary_key(name scope, name table, uint64_t primary_key) {
+      const std::size_t zero_extension_size = 0;
+      b1::chain_kv::bytes composite_key = detail::prepare_composite_key(scope, table, sizeof(uint64_t), key_type::primary, zero_extension_size);
+      b1::chain_kv::append_key(composite_key, primary_key);
       return composite_key;
    }
 
-   void get_primary_key(const b1::chain_kv::bytes& composite_key, name& scope, name& table, uint64_t& db_key) {
+   bool get_primary_key(const b1::chain_kv::bytes& composite_key, name& scope, name& table, uint64_t& primary_key) {
       b1::chain_kv::bytes::const_iterator composite_loc;
       key_type kt = key_type::sec_i64;
       std::tie(scope, table, composite_loc, kt) = detail::extract_from_composite_key(composite_key.cbegin(), composite_key.cend());
-      EOS_ASSERT(kt == key_type::primary, bad_composite_key_exception,
-                 "DB intrinsic key-value store composite key is malformed, it is supposed to be a primary key, "
-                 "but it is a: " + detail::to_string(kt));
-      EOS_ASSERT(b1::chain_kv::extract_key(composite_loc, composite_key.cend(), db_key), bad_composite_key_exception,
+      if (kt != key_type::primary) {
+         return false;
+      }
+      EOS_ASSERT(b1::chain_kv::extract_key(composite_loc, composite_key.cend(), primary_key), bad_composite_key_exception,
                  "DB intrinsic key-value store composite key is malformed, it is supposed to have a primary key");
+      EOS_ASSERT(composite_loc == composite_key.cend(), bad_composite_key_exception,
+                 "DB intrinsic key-value store composite key is malformed, it has extra data after the primary key");
+      return true;
+   }
+
+   b1::chain_kv::bytes create_table_key(name scope, name table) {
+      // table key ends with the type, so just reuse method
+      return create_prefix_type_key<key_type::table>(scope, table);
+   }
+
+   eosio::session::shared_bytes create_table_key(const eosio::session::shared_bytes& prefix_key) {
+      const std::size_t full_prefix_size = db_key_value_format::detail::prefix_size<eosio::session::shared_bytes>();
+      EOS_ASSERT(prefix_key.size() >= full_prefix_size, bad_composite_key_exception,
+                 "DB intrinsic key-value store composite key is malformed, create_table_key expects to be given a full "
+                 "key that has a full prefix leading up to the type");
+      const auto new_kt = static_cast<char>(key_type::table);
+      const char* new_kt_ptr = &new_kt;
+      // already have everything in char format/order, so just need to assemble it
+      return detail::assemble<2, eosio::session::shared_bytes>({rocksdb::Slice{prefix_key.data(), full_prefix_size},
```

Comment:
We should get rid of the dependency of rocksdb in this file.

Reply:
I was planning on cleaning out the old code that used Slices, but figured I would use the same type that it is using throughout the cpp and hpp and then clean them all up together.